worker/modelupgrader/worker.go

```diff
@@ -0,0 +1,236 @@
+	}
+	if config.GateUnlocker == nil {
+		return errors.NotValidf("nil GateUnlocker")
+	}
+	if config.ControllerTag == (names.ControllerTag{}) {
+		return errors.NotValidf("empty ControllerTag")
+	}
+	if config.ModelTag == (names.ModelTag{}) {
+		return errors.NotValidf("empty ModelTag")
+	}
+	return nil
+}
+
+// NewWorker returns a worker that ensures that environ/provider schema upgrades
+// are run when the model is first loaded by a controller of a new version. The
+// worker either runs the upgrades or waits for another controller unit to run
+// them, depending on the configuration.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	targetVersion, err := config.Facade.ModelTargetEnvironVersion(config.ModelTag)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if config.Environ != nil {
+		return newUpgradeWorker(config, targetVersion)
+	}
+	return newWaitWorker(config, targetVersion)
+}
+
+// newWaitWorker returns a worker that waits for the controller leader to run
+// the upgrade steps and update the model's environ version, and then unlocks
+// the gate.
+func newWaitWorker(config Config, targetVersion int) (worker.Worker, error) {
+	watcher, err := config.Facade.WatchModelEnvironVersion(config.ModelTag)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	ww := waitWorker{
+		watcher:       watcher,
+		facade:        config.Facade,
+		modelTag:      config.ModelTag,
+		gate:          config.GateUnlocker,
+		targetVersion: targetVersion,
+	}
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &ww.catacomb,
+		Init: []worker.Worker{watcher},
+		Work: ww.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &ww, nil
+}
+
+type waitWorker struct {
+	catacomb      catacomb.Catacomb
+	watcher       watcher.NotifyWatcher
+	facade        Facade
+	modelTag      names.ModelTag
+	gate          gate.Unlocker
+	targetVersion int
+}
+
+func (ww *waitWorker) Kill() {
+	ww.catacomb.Kill(nil)
+}
+
+func (ww *waitWorker) Wait() error {
+	return ww.catacomb.Wait()
+}
+
+func (ww *waitWorker) loop() error {
+	for {
+		select {
+		case <-ww.catacomb.Dying():
+			return ww.catacomb.ErrDying()
+		case _, ok := <-ww.watcher.Changes():
+			if !ok {
+				return ww.catacomb.ErrDying()
+			}
+			currentVersion, err := ww.facade.ModelEnvironVersion(ww.modelTag)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if currentVersion >= ww.targetVersion {
+				ww.gate.Unlock()
+				return nil
+			}
+		}
+	}
+}
+
+// newUpgradeWorker returns a worker that runs the upgrade steps, updates the
+// model's environ version, and unlocks the gate.
+func newUpgradeWorker(config Config, targetVersion int) (worker.Worker, error) {
+	currentVersion, err := config.Facade.ModelEnvironVersion(config.ModelTag)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return jujuworker.NewSimpleWorker(func(<-chan struct{}) error {
```

Comment:
Only if we make the upgrade steps interruptible. They're not supposed to be long-running things anyway, so I don't think it's necessary.